src/crab/ebpf_domain.cpp

```diff
@@ -2114,8 +2119,57 @@ void ebpf_domain_t::do_mem_store(const Mem& b, Type val_type, SValue val_svalue,
     });
 }
 
-void ebpf_domain_t::operator()(const LockAdd& a) {
-    // nothing to do here
+void ebpf_domain_t::operator()(const Atomic& a) {
+    if (m_inv.is_bottom())
+        return;
+    if (!m_inv.entail(type_is_shared(reg_pack(a.access.basereg))) ||
+        !m_inv.entail(type_is_number(reg_pack(a.valreg)))) {
+        return;
+    }
+
+    // Fetch the current value into the R11 pseudo-register.
+    Reg r11{R11_ATOMIC_SCRATCH};
+    Mem mem = {.access = a.access, .value = r11, .is_load = true};
+    (*this)(mem);
+
+    // Compute the new value in R11.
+    Bin bin {.dst = r11, .v = a.valreg, .is64 = (a.access.width == 64), .lddw = false};
+    switch ((AtomicOp)((uint32_t)a.op & (uint32_t)AtomicOp::INST_ATOMIC_BASE_MASK)) {
+    case AtomicOp::INST_ATOMIC_ADD: bin.op = Bin::Op::ADD; break;
+    case AtomicOp::INST_ATOMIC_OR: bin.op = Bin::Op::OR; break;
+    case AtomicOp::INST_ATOMIC_AND: bin.op = Bin::Op::AND; break;
+    case AtomicOp::INST_ATOMIC_XOR: bin.op = Bin::Op::XOR; break;
+    case AtomicOp::INST_ATOMIC_XCHG_BASE:
+    case AtomicOp::INST_ATOMIC_CMPXCHG_BASE: bin.op = Bin::Op::MOV; break;
+    default: throw std::exception();
+    }
+    (*this)(bin);
+
+    if (a.op == AtomicOp::INST_ATOMIC_CMPXCHG) {
+        // For CMPXCHG, store the original value in r0 if and only if the value matches
+        // what is in the src register.  Until we track shared memory values,
```

Comment:
Is it theoretically possible to track shared memory values?

Reply:
Yes, just like we track stack memory values.  The problem is that there could be a huge amount of shared memory space, unlike stack space.

Reply:
Hmm, on the other hand, if it's shared, you don't know when it might be modified by something else so in that sense I guess you can't reliably track the value.

Reply:
Would you like this updated to not assume shared memory is ever trackable?

Reply:
I understood shared as volatile, so writes are no-ops and reads are havocs.

Reply:
Updated to not assume shared memory values are ever trackable.